src/pds_doi_service/api/__main__.py

```diff
@@ -18,6 +20,118 @@
 
 logging.basicConfig(level=logging.INFO)
 
+# We create the connexion app here so we can access underlying Flask decorators
+app = connexion.App(__name__, specification_dir='swagger/')
+# We also add an initialization flag to ensure that calls to init_app() only
+# add the swagger api once
+app.initialized = False
+
+
+class InvalidReferrer(Exception):
+    """Raised when the referrer check fails."""
+    status_code = 401
+
+    def __init__(self, message, status_code=None):
+        Exception.__init__(self)
+        self.message = message
+
+        if status_code is not None:
+            self.status_code = status_code
+
+    def to_dict(self):
+        rv = {'message': self.message, 'status_code': self.status_code}
+        return rv
+
+
+@app.app.errorhandler(InvalidReferrer)
+def handle_invalid_usage(error):
+    """Registers a handler for InvalidReferrer exceptions caught by Flask"""
+    response = jsonify(error.to_dict())
+    response.status_code = error.status_code
+    return response
+
+
+@app.app.before_request
+def _check_referrer():
+    """
+    Checks the referrer header field of the current request to ensure it was
+    sent by a valid endpoint. The list of valid referrers is defined by the
+    DOI service INI config. If not valid referrers are defined by the config,
+    this function will allow all requests through.
+
+    Raises
+    ------
+    InvalidReferrer
+        If no referrer is provided with the request, or the referrer is not
+        within the list of valid hosts.
+
+    Returns
+    -------
+    None if the referrer is valid (meaning the request should continue).
+
+    """
+    logger = logging.getLogger(__name__)
+    config = DOIConfigUtil().get_config()
+
+    valid_referrers = config.get('OTHER', 'api_valid_referrers')
+    valid_referrers = list(map(str.strip, valid_referrers.split(',')))
```

Comment:
How about using a `set` here instead of a `list`? It would remove duplicates if, for example, someone did:
```
api_valid_referrers = localhost, 172.16.16.27, localhost, 0.0.0.0, localhost
```
and it'd give you a good speed boost (lookup of `in` with set membership is faster than `in` with list).

Reply:
@nutjob4life that is good to know, thanks